docs/_posts/2022-10-31-luau-semantic-subtyping.md

```diff
@@ -0,0 +1,239 @@
+
+Once types are normalized, it is straightforward to check semantic subtyping.
+
+Every type can be normalized (sigh, with some technical restrictions around generic type packs). The important steps are:
+
+* removing intersections of mismatched primitives, e.g. `number & bool` is replaced by `never`, and
+* removing unions of functions, e.g. `((number?) -> number) | ((string?) -> string)` is replaced by `(nil) -> (number | string)`.
+
+For example, normalizing `(number?) & (string?)` removes `number & string`, so all that is left is `nil`.
+
+Our first attempt at implementing type normalization applied it liberally, but this resulted in dreadful performance (complex code went from typechecking in less than a minute to running overnight). The reason for this is annoyingly simple: there is an optimization in Luau’s subtyping algorithm to handle reflexivity (`T` is a subtype of `T`) that performs a cheap pointer equality check. Type normalization can convert pointer-identical types into semantically-equivalent (but not pointer-identical) types, which significantly degrades performance.
+
+Because of these performance issues, we still use syntactic subtyping as our first check for subtyping, and only perform type normalization if the syntactic algorithm fails. This is sound, because syntactic subtyping is a conservative approximation to semantic subtyping.
+
+## Pragmatic semantic subtyping
+
+Off-the-shelf semantic subtyping is slightly different from what is implemented in Luau, because it requires models to be *set-theoretic*, which requires that inhabitants of function types “act like functions”. There are two reasons why we drop this requirement.
+
+**Firstly**, we normalize function types to an intersection of functions, for example a horrible mess of unions and intersections of functions:
+```
+(number? -> number?) | ((number -> number) & (string? -> string?))
+```
+normalizes to an overloaded function:
+```
+(number -> number?) & (nil -> (number | string)?)
+```
+Set-theoretic semantic subtyping does not support this normalization, and instead normalizes functions to *disjunctive normal form* (unions of intersections of functions). We do not do this for ergonomic reasons: overloaded functions are idiomatic in Luau, but DNF is not, and we do not want to present users with such non-idiomatic types.
+
+Our normalization relies on rewriting away unions of function types:
+```
+(A -> B) | (C -> D)   →   (A & C) -> (B | D) 
+```
+This normalization is sound in our model, but not in set-theoretic models.
+
+**Secondly**, in Luau, the type of a function application `f(x)` is `B` if `f` has type `A -> B` and `x` has type `A`. Unexpectedly, this is not always true in set-theoretic models, due to uninhabited types. In set-theoretic models, if `x` has type `never` then `f(x)` has type `never`. We do not want to burden users with the idea that function application has a special corner case, especially since that corner case can only arise in dead code.
```

Comment:
```suggestion
**Secondly**, in Luau, the type of a function application `f(x)` is `B` if `f` has type `(A) -> B` and `x` has type `A`. Unexpectedly, this is not always true in set-theoretic models, due to uninhabited types. In set-theoretic models, if `x` has type `never` then `f(x)` has type `never`. We do not want to burden users with the idea that function application has a special corner case, especially since that corner case can only arise in dead code.
```